File: ChainBridge/api/schemas/governance_health.py
"""
Governance Health Schemas — PAC-CODY-P01-GOVERNANCE-HEALTH-BACKEND-AGGREGATION-01

Pydantic models for governance health API responses.
All models are READ-ONLY — no mutation operations.

Authority: CODY (GID-01)
Dispatch: PAC-BENSON-EXEC-P61
"""

from datetime import datetime
from enum import Enum
from typing import List, Optional

from pydantic import BaseModel, Field


# =============================================================================
# ENUMS
# =============================================================================


class ArtifactStatus(str, Enum):
    """Status of governance artifacts in settlement flow."""
    PENDING = "PENDING"
    ACTIVE = "ACTIVE"
    AWAITING_REVIEW = "AWAITING_REVIEW"
    FINALIZED = "FINALIZED"
    REJECTED = "REJECTED"
    BLOCKED = "BLOCKED"


class SettlementStage(str, Enum):
    """Stage in the PAC → BER → PDO → WRAP settlement flow."""
    PAC_DISPATCH = "PAC_DISPATCH"
    AGENT_EXECUTION = "AGENT_EXECUTION"
    BER_GENERATION = "BER_GENERATION"
    HUMAN_REVIEW = "HUMAN_REVIEW"
    PDO_FINALIZATION = "PDO_FINALIZATION"
    WRAP_GENERATION = "WRAP_GENERATION"
    WRAP_ACCEPTED = "WRAP_ACCEPTED"
    LEDGER_COMMIT = "LEDGER_COMMIT"


class ChainStatus(str, Enum):
    """Status of a settlement chain."""
    IN_PROGRESS = "IN_PROGRESS"
    COMPLETED = "COMPLETED"
    BLOCKED = "BLOCKED"
    REJECTED = "REJECTED"


class LedgerIntegrity(str, Enum):
    """Health status of the governance ledger."""
    HEALTHY = "HEALTHY"
    DEGRADED = "DEGRADED"
    CRITICAL = "CRITICAL"


class ComplianceFramework(str, Enum):
    """Enterprise compliance frameworks."""
    SOX = "SOX"
    SOC2 = "SOC2"
    NIST_CSF = "NIST_CSF"
    ISO_27001 = "ISO_27001"


class GovernanceArtifact(str, Enum):
    """Types of governance artifacts."""
    PAC = "PAC"
    BER = "BER"
    PDO = "PDO"
    WRAP = "WRAP"
    LEDGER = "LEDGER"


# =============================================================================
# RESPONSE MODELS
# =============================================================================


class GovernanceHealthMetrics(BaseModel):
    """Governance health metrics for dashboard."""
    
    # PAC metrics
    total_pacs: int = Field(..., description="Total PACs in ledger")
    active_pacs: int = Field(..., description="PACs currently in execution")
    blocked_pacs: int = Field(..., description="PACs blocked for correction")
    positive_closures: int = Field(..., description="PACs with positive closure")
    
    # BER metrics
    total_bers: int = Field(..., description="Total BERs generated")
    pending_bers: int = Field(..., description="BERs awaiting review")
    approved_bers: int = Field(..., description="BERs approved for PDO")
    
    # PDO metrics
    total_pdos: int = Field(..., description="Total PDOs created")
    finalized_pdos: int = Field(..., description="PDOs finalized")
    
    # WRAP metrics
    total_wraps: int = Field(..., description="Total WRAPs generated")
    accepted_wraps: int = Field(..., description="WRAPs accepted to ledger")
    
    # Settlement metrics
    settlement_rate: float = Field(..., description="Percentage of PACs reaching WRAP_ACCEPTED")
    avg_settlement_time_ms: int = Field(..., description="Average settlement time in milliseconds")
    pending_settlements: int = Field(..., description="Settlements in progress")
    
    # Health indicators
    ledger_integrity: LedgerIntegrity = Field(..., description="Ledger health status")
    last_ledger_sync: datetime = Field(..., description="Last ledger sync timestamp")
    sequence_gaps: int = Field(..., description="Number of sequence gaps detected")


class SettlementFlowNode(BaseModel):
    """A node in the settlement flow."""
    stage: SettlementStage = Field(..., description="Settlement stage")
    status: ArtifactStatus = Field(..., description="Status at this stage")
    timestamp: Optional[datetime] = Field(None, description="When stage was reached")
    artifact_id: Optional[str] = Field(None, description="Associated artifact ID")
    authority: Optional[str] = Field(None, description="Authority at this stage")
    details: Optional[str] = Field(None, description="Additional details")


class SettlementChain(BaseModel):
    """A settlement chain from PAC to WRAP."""
    chain_id: str = Field(..., description="Unique chain identifier")
    pac_id: str = Field(..., description="PAC ID")
    ber_id: Optional[str] = Field(None, description="BER ID if generated")
    pdo_id: Optional[str] = Field(None, description="PDO ID if finalized")
    wrap_id: Optional[str] = Field(None, description="WRAP ID if accepted")
    current_stage: SettlementStage = Field(..., description="Current settlement stage")
    status: ChainStatus = Field(..., description="Overall chain status")
    started_at: datetime = Field(..., description="When chain started")
    completed_at: Optional[datetime] = Field(None, description="When chain completed")
    nodes: List[SettlementFlowNode] = Field(..., description="Flow nodes")
    agent_gid: str = Field(..., description="Agent GID")
    agent_name: str = Field(..., description="Agent name")


class EnterpriseMapping(BaseModel):
    """Mapping of governance artifact to compliance framework."""
    framework: ComplianceFramework = Field(..., description="Compliance framework")
    control: str = Field(..., description="Control identifier")
    description: str = Field(..., description="Control description")
    artifact: GovernanceArtifact = Field(..., description="Mapped artifact type")


class FrameworkCoverage(BaseModel):
    """Coverage percentage per framework."""
    sox: float = Field(..., description="SOX coverage percentage")
    soc2: float = Field(..., description="SOC 2 coverage percentage")
    nist: float = Field(..., description="NIST CSF coverage percentage")
    iso27001: float = Field(..., description="ISO 27001 coverage percentage")


class EnterpriseComplianceSummary(BaseModel):
    """Enterprise compliance summary."""
    mappings: List[EnterpriseMapping] = Field(..., description="All compliance mappings")
    last_audit_date: Optional[str] = Field(None, description="Last audit date")
    compliance_score: float = Field(..., description="Overall compliance score (0-100)")
    framework_coverage: FrameworkCoverage = Field(..., description="Coverage per framework")


# =============================================================================
# API RESPONSE WRAPPERS
# =============================================================================


class GovernanceHealthResponse(BaseModel):
    """Response wrapper for governance health endpoint."""
    success: bool = Field(True, description="Request success status")
    data: GovernanceHealthMetrics = Field(..., description="Health metrics")
    timestamp: datetime = Field(..., description="Response timestamp")
    source: str = Field("GOVERNANCE_LEDGER", description="Data source")


class SettlementChainsResponse(BaseModel):
    """Response wrapper for settlement chains endpoint."""
    success: bool = Field(True, description="Request success status")
    data: List[SettlementChain] = Field(..., description="Settlement chains")
    total: int = Field(..., description="Total chains")
    timestamp: datetime = Field(..., description="Response timestamp")


class ComplianceSummaryResponse(BaseModel):
    """Response wrapper for compliance summary endpoint."""
    success: bool = Field(True, description="Request success status")
    data: EnterpriseComplianceSummary = Field(..., description="Compliance summary")
    timestamp: datetime = Field(..., description="Response timestamp")
    doctrine_version: str = Field("1.1.0", description="Governance Doctrine version")


class ErrorResponse(BaseModel):
    """Error response model."""
    success: bool = Field(False, description="Request failed")
    error_code: str = Field(..., description="Error code")
    message: str = Field(..., description="Error message")
    timestamp: datetime = Field(..., description="Error timestamp")

File: ChainBridge/chainboard-ui/src/types/governanceHealth.ts
/**
 * Governance Health Types — PAC-SONNY-P01-GOVERNANCE-HEALTH-DASHBOARD-01
 *
 * Types for the Governance Health Dashboard visualization.
 * All types are READ-ONLY and derived from governance artifacts.
 *
 * Settlement Chain: PAC → BER → PDO → WRAP
 *
 * @see PAC-SONNY-P01-GOVERNANCE-HEALTH-DASHBOARD-01
 * @see GOVERNANCE_DOCTRINE_V1.1
 */

/**
 * Settlement artifact status in the governance flow.
 */
export type ArtifactStatus =
  | 'PENDING'
  | 'ACTIVE'
  | 'AWAITING_REVIEW'
  | 'FINALIZED'
  | 'REJECTED'
  | 'BLOCKED';

/**
 * Settlement stage in the PAC → BER → PDO → WRAP flow.
 */
export type SettlementStage =
  | 'PAC_DISPATCH'
  | 'AGENT_EXECUTION'
  | 'BER_GENERATION'
  | 'HUMAN_REVIEW'
  | 'PDO_FINALIZATION'
  | 'WRAP_GENERATION'
  | 'WRAP_ACCEPTED'
  | 'LEDGER_COMMIT';

/**
 * Enterprise compliance framework mapping.
 */
export interface EnterpriseMapping {
  framework: 'SOX' | 'SOC2' | 'NIST_CSF' | 'ISO_27001';
  control: string;
  description: string;
  artifact: 'PAC' | 'BER' | 'PDO' | 'WRAP' | 'LEDGER';
}

/**
 * Settlement flow node for visualization.
 */
export interface SettlementFlowNode {
  stage: SettlementStage;
  status: ArtifactStatus;
  timestamp?: string;
  artifactId?: string;
  authority?: string;
  details?: string;
}

/**
 * Governance health metrics for dashboard cards.
 */
export interface GovernanceHealthMetrics {
  // PAC metrics
  totalPacs: number;
  activePacs: number;
  blockedPacs: number;
  positiveClosures: number;

  // BER metrics
  totalBers: number;
  pendingBers: number;
  approvedBers: number;

  // PDO metrics
  totalPdos: number;
  finalizedPdos: number;

  // WRAP metrics
  totalWraps: number;
  acceptedWraps: number;

  // Settlement metrics
  settlementRate: number; // percentage of PACs reaching WRAP_ACCEPTED
  avgSettlementTimeMs: number;
  pendingSettlements: number;

  // Health indicators
  ledgerIntegrity: 'HEALTHY' | 'DEGRADED' | 'CRITICAL';
  lastLedgerSync: string;
  sequenceGaps: number;
}

/**
 * Artifact in the settlement timeline.
 */
export interface SettlementArtifact {
  id: string;
  type: 'PAC' | 'BER' | 'PDO' | 'WRAP';
  status: ArtifactStatus;
  createdAt: string;
  updatedAt: string;
  agentGid: string;
  agentName: string;
  agentColor: string;
  linkedArtifacts: {
    pacId?: string;
    berId?: string;
    pdoId?: string;
    wrapId?: string;
  };
}

/**
 * Settlement chain representing a complete PAC → WRAP flow.
 */
export interface SettlementChain {
  chainId: string;
  pacId: string;
  berId?: string;
  pdoId?: string;
  wrapId?: string;
  currentStage: SettlementStage;
  status: 'IN_PROGRESS' | 'COMPLETED' | 'BLOCKED' | 'REJECTED';
  startedAt: string;
  completedAt?: string;
  nodes: SettlementFlowNode[];
}

/**
 * Enterprise compliance summary for the dashboard.
 */
export interface EnterpriseComplianceSummary {
  mappings: EnterpriseMapping[];
  lastAuditDate?: string;
  complianceScore: number; // 0-100
  frameworkCoverage: {
    sox: number;
    soc2: number;
    nist: number;
    iso27001: number;
  };
}

/**
 * Governance health dashboard state.
 */
export interface GovernanceHealthState {
  metrics: GovernanceHealthMetrics;
  recentChains: SettlementChain[];
  complianceSummary: EnterpriseComplianceSummary;
  isLoading: boolean;
  error: Error | null;
  lastUpdated: string;
}

/**
 * Agent color mapping for consistent UI.
 */
export const AGENT_COLORS: Record<string, string> = {
  TEAL: 'bg-teal-500',
  BLUE: 'bg-blue-500',
  YELLOW: 'bg-yellow-500',
  PURPLE: 'bg-purple-500',
  CYAN: 'bg-cyan-500',
  DARK_RED: 'bg-red-700',
  GREEN: 'bg-green-500',
  WHITE: 'bg-slate-200',
  PINK: 'bg-pink-500',
  MAGENTA: 'bg-pink-600',
  CRIMSON: 'bg-red-600',
  ORANGE: 'bg-orange-500',
};

/**
 * Stage display configuration.
 */
export const STAGE_CONFIG: Record<SettlementStage, { label: string; icon: string }> = {
  PAC_DISPATCH: { label: 'PAC Dispatch', icon: '📋' },
  AGENT_EXECUTION: { label: 'Agent Execution', icon: '⚙️' },
  BER_GENERATION: { label: 'BER Generation', icon: '📊' },
  HUMAN_REVIEW: { label: 'Human Review', icon: '👁️' },
  PDO_FINALIZATION: { label: 'PDO Finalization', icon: '🔐' },
  WRAP_GENERATION: { label: 'WRAP Generation', icon: '📦' },
  WRAP_ACCEPTED: { label: 'WRAP Accepted', icon: '✅' },
  LEDGER_COMMIT: { label: 'Ledger Commit', icon: '📖' },
};

/**
 * Status color mapping.
 */
export const STATUS_COLORS: Record<ArtifactStatus, string> = {
  PENDING: 'text-slate-400 bg-slate-500/10',
  ACTIVE: 'text-blue-400 bg-blue-500/10',
  AWAITING_REVIEW: 'text-amber-400 bg-amber-500/10',
  FINALIZED: 'text-emerald-400 bg-emerald-500/10',
  REJECTED: 'text-rose-400 bg-rose-500/10',
  BLOCKED: 'text-red-400 bg-red-500/10',
};
